File: examples/android-compose-host/app/src/main/java/io/github/ikaros/rustplayersdk/example/androidcomposehost/ExamplePlayerSections.kt
package io.github.ikaros.vesper.example.androidcomposehost

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.BrightnessAuto
import androidx.compose.material.icons.rounded.DarkMode
import androidx.compose.material.icons.rounded.LightMode
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.layout.Layout
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.max

@Composable
internal fun ExamplePlayerHeader(
    sourceLabel: String,
    subtitle: String,
    palette: ExampleHostPalette,
) {
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Text(
            text = "Vesper",
            style = MaterialTheme.typography.headlineLarge.copy(
                color = palette.title,
                fontWeight = FontWeight.Black,
                letterSpacing = (-1.2).sp,
            ),
        )
        Text(
            text = sourceLabel,
            style = MaterialTheme.typography.titleSmall.copy(
                color = palette.title,
                fontWeight = FontWeight.SemiBold,
            ),
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
        Text(
            text = subtitle,
            style = MaterialTheme.typography.bodyMedium.copy(color = palette.body),
            maxLines = 2,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

@Composable
internal fun ExampleSourceSection(
    palette: ExampleHostPalette,
    themeMode: ExampleThemeMode,
    remoteStreamUrl: String,
    onThemeModeChange: (ExampleThemeMode) -> Unit,
    onRemoteStreamUrlChange: (String) -> Unit,
    onPickVideo: () -> Unit,
    onUseHlsDemo: () -> Unit,
    onUseDashDemo: () -> Unit,
    onOpenRemote: () -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(palette.sectionBackground, RoundedCornerShape(24.dp))
            .border(1.dp, palette.sectionStroke, RoundedCornerShape(24.dp))
            .padding(18.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Text(
            text = stringResource(R.string.example_sources_title),
            style = MaterialTheme.typography.titleMedium.copy(
                color = palette.title,
                fontWeight = FontWeight.Bold,
            ),
        )
        Text(
            text = stringResource(R.string.example_sources_subtitle),
            style = MaterialTheme.typography.bodySmall.copy(color = palette.body),
        )

        Row(
            modifier = Modifier.horizontalScroll(rememberScrollState()),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            OutlinedButton(onClick = onPickVideo) {
                Text(stringResource(R.string.example_sources_pick_video))
            }
            OutlinedButton(onClick = onUseHlsDemo) {
                Text(stringResource(R.string.example_sources_hls_demo))
            }
            OutlinedButton(onClick = onUseDashDemo) {
                Text(stringResource(R.string.example_sources_dash_demo))
            }
        }

        OutlinedTextField(
            value = remoteStreamUrl,
            onValueChange = onRemoteStreamUrlChange,
            modifier = Modifier.fillMaxWidth(),
            label = { Text(stringResource(R.string.example_sources_remote_stream_url)) },
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Uri),
            singleLine = true,
        )

        Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
            Text(
                text = stringResource(R.string.example_sources_theme),
                style = MaterialTheme.typography.labelLarge.copy(
                    color = palette.title,
                    fontWeight = FontWeight.SemiBold,
                ),
            )
            Row(
                modifier = Modifier.horizontalScroll(rememberScrollState()),
                horizontalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                ThemeModeChip(
                    icon = Icons.Rounded.BrightnessAuto,
                    label = stringResource(ExampleThemeMode.System.titleRes),
                    selected = themeMode == ExampleThemeMode.System,
                    onClick = { onThemeModeChange(ExampleThemeMode.System) },
                )
                ThemeModeChip(
                    icon = Icons.Rounded.LightMode,
                    label = stringResource(ExampleThemeMode.Light.titleRes),
                    selected = themeMode == ExampleThemeMode.Light,
                    onClick = { onThemeModeChange(ExampleThemeMode.Light) },
                )
                ThemeModeChip(
                    icon = Icons.Rounded.DarkMode,
                    label = stringResource(ExampleThemeMode.Dark.titleRes),
                    selected = themeMode == ExampleThemeMode.Dark,
                    onClick = { onThemeModeChange(ExampleThemeMode.Dark) },
                )
            }
        }

        Button(
            onClick = onOpenRemote,
            colors = ButtonDefaults.buttonColors(
                containerColor = palette.primaryAction,
                contentColor = Color.White,
            ),
        ) {
            Text(stringResource(R.string.example_sources_open_remote_url))
        }
    }
}

@Composable
internal fun ExampleResilienceSection(
    palette: ExampleHostPalette,
    selectedProfile: ExampleResilienceProfile,
    isApplyingProfile: Boolean,
    onApplyProfile: (ExampleResilienceProfile) -> Unit,
) {
    val policy = selectedProfile.policy
    ExampleSectionShell(
        palette = palette,
        title = stringResource(R.string.example_resilience_title),
        subtitle = stringResource(R.string.example_resilience_subtitle),
    ) {
        Column(verticalArrangement = Arrangement.spacedBy(14.dp)) {
            AdaptiveChipWrap(
                modifier = Modifier.fillMaxWidth(),
                horizontalSpacing = 10.dp,
                verticalSpacing = 10.dp,
            ) {
                ExampleResilienceProfile.values().forEach { profile ->
                    SelectionChip(
                        label = stringResource(profile.titleRes),
                        selected = profile == selectedProfile,
                        onClick = { onApplyProfile(profile) },
                    )
                }
            }

            Text(
                text = stringResource(selectedProfile.subtitleRes),
                style = MaterialTheme.typography.bodyMedium.copy(
                    color = palette.body,
                    lineHeight = 22.sp,
                ),
            )

            if (isApplyingProfile) {
                ExampleStatusPill(
                    label = stringResource(R.string.example_resilience_applying),
                    palette = palette,
                )
            }

            Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                ExampleFactRow(
                    label = stringResource(R.string.example_resilience_fact_buffering),
                    value = resilienceBufferingValue(policy.buffering),
                    palette = palette,
                )
                ExampleFactRow(
                    label = stringResource(R.string.example_resilience_fact_retry),
                    value = resilienceRetryValue(policy.retry),
                    palette = palette,
                )
                ExampleFactRow(
                    label = stringResource(R.string.example_resilience_fact_cache),
                    value = resilienceCacheValue(policy.cache),
                    palette = palette,
                )
            }
        }
    }
}

@Composable
internal fun ThemeModeChip(
    icon: ImageVector,
    label: String,
    selected: Boolean,
    onClick: () -> Unit,
) {
    TextButton(
        onClick = onClick,
        colors = ButtonDefaults.textButtonColors(
            contentColor = if (selected) Color.White else MaterialTheme.colorScheme.onSurface,
        ),
        contentPadding = PaddingValues(horizontal = 12.dp, vertical = 8.dp),
        modifier = Modifier
            .heightIn(min = 38.dp)
            .background(
                if (selected) {
                    MaterialTheme.colorScheme.primary
                } else {
                    MaterialTheme.colorScheme.surface.copy(alpha = 0.72f)
                },
                RoundedCornerShape(999.dp),
            ),
    ) {
        Icon(
            imageVector = icon,
            contentDescription = null,
            modifier = Modifier.size(16.dp),
        )
        Spacer(modifier = Modifier.width(6.dp))
        Text(label, maxLines = 1)
    }
}

@Composable
private fun SelectionChip(
    label: String,
    selected: Boolean,
    onClick: () -> Unit,
) {
    TextButton(
        onClick = onClick,
        colors = ButtonDefaults.textButtonColors(
            contentColor = if (selected) Color.White else MaterialTheme.colorScheme.onSurface,
        ),
        contentPadding = PaddingValues(horizontal = 14.dp, vertical = 8.dp),
        modifier = Modifier
            .heightIn(min = 38.dp)
            .background(
                if (selected) {
                    MaterialTheme.colorScheme.primary
                } else {
                    MaterialTheme.colorScheme.surface.copy(alpha = 0.72f)
                },
                RoundedCornerShape(999.dp),
            ),
    ) {
        Text(label, maxLines = 1)
    }
}

@Composable
private fun AdaptiveChipWrap(
    modifier: Modifier = Modifier,
    horizontalSpacing: androidx.compose.ui.unit.Dp,
    verticalSpacing: androidx.compose.ui.unit.Dp,
    content: @Composable () -> Unit,
) {
    Layout(
        modifier = modifier,
        content = content,
    ) { measurables, constraints ->
        val horizontalSpacingPx = horizontalSpacing.roundToPx()
        val verticalSpacingPx = verticalSpacing.roundToPx()
        val maxWidth = constraints.maxWidth.takeIf { it < Int.MAX_VALUE } ?: Int.MAX_VALUE
        val placeables = measurables.map { measurable ->
            measurable.measure(constraints.copy(minWidth = 0, minHeight = 0))
        }

        data class PositionedPlaceable(
            val placeable: androidx.compose.ui.layout.Placeable,
            val x: Int,
            val y: Int,
        )

        val positionedPlaceables = mutableListOf<PositionedPlaceable>()
        var currentX = 0
        var currentY = 0
        var currentRowHeight = 0
        var contentWidth = 0

        placeables.forEach { placeable ->
            val shouldWrap =
                currentX > 0 &&
                    currentX + placeable.width > maxWidth
            if (shouldWrap) {
                currentX = 0
                currentY += currentRowHeight + verticalSpacingPx
                currentRowHeight = 0
            }

            positionedPlaceables +=
                PositionedPlaceable(
                    placeable = placeable,
                    x = currentX,
                    y = currentY,
                )
            currentX += placeable.width + horizontalSpacingPx
            currentRowHeight = max(currentRowHeight, placeable.height)
            contentWidth = max(contentWidth, currentX - horizontalSpacingPx)
        }

        val widthCandidate =
            if (constraints.maxWidth < Int.MAX_VALUE) {
                constraints.maxWidth
            } else {
                contentWidth
            }
        val heightCandidate =
            if (positionedPlaceables.isEmpty()) {
                0
            } else {
                currentY + currentRowHeight
            }
        val layoutWidth = widthCandidate.coerceIn(constraints.minWidth, constraints.maxWidth)
        val layoutHeight = heightCandidate.coerceIn(constraints.minHeight, constraints.maxHeight)

        layout(layoutWidth, layoutHeight) {
            positionedPlaceables.forEach { positioned ->
                positioned.placeable.placeRelative(positioned.x, positioned.y)
            }
        }
    }
}

@Composable
private fun ExampleSectionShell(
    palette: ExampleHostPalette,
    title: String,
    subtitle: String,
    content: @Composable () -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(palette.sectionBackground, RoundedCornerShape(24.dp))
            .border(1.dp, palette.sectionStroke, RoundedCornerShape(24.dp))
            .padding(18.dp),
    ) {
        Text(
            text = title,
            style = MaterialTheme.typography.titleMedium.copy(
                color = palette.title,
                fontWeight = FontWeight.Bold,
            ),
        )
        Spacer(modifier = Modifier.size(8.dp))
        Text(
            text = subtitle,
            style = MaterialTheme.typography.bodySmall.copy(
                color = palette.body,
                lineHeight = 20.sp,
            ),
        )
        Spacer(modifier = Modifier.size(14.dp))
        Box(
            modifier = Modifier
                .width(42.dp)
                .height(4.dp)
                .background(Color(0xFF172033), RoundedCornerShape(999.dp)),
        )
        Spacer(modifier = Modifier.size(16.dp))
        content()
    }
}

@Composable
private fun ExampleFactRow(
    label: String,
    value: String,
    palette: ExampleHostPalette,
) {
    Row(modifier = Modifier.padding(vertical = 6.dp)) {
        Text(
            text = label,
            modifier = Modifier.width(92.dp),
            style = MaterialTheme.typography.bodyMedium.copy(color = palette.body),
        )
        Spacer(modifier = Modifier.width(10.dp))
        Text(
            text = value,
            modifier = Modifier.weight(1f),
            style = MaterialTheme.typography.bodyMedium.copy(
                color = palette.title,
                fontWeight = FontWeight.SemiBold,
            ),
        )
    }
}

@Composable
private fun ExampleStatusPill(
    label: String,
    palette: ExampleHostPalette,
) {
    Text(
        text = label,
        modifier = Modifier
            .background(
                palette.primaryAction.copy(alpha = 0.12f),
                RoundedCornerShape(999.dp),
            )
            .padding(horizontal = 12.dp, vertical = 7.dp),
        style = MaterialTheme.typography.labelMedium.copy(
            color = palette.primaryAction,
            fontWeight = FontWeight.SemiBold,
        ),
    )
}

File: examples/flutter-host/lib/src/player/example_player_sections.dart
import 'package:flutter/material.dart';
import 'package:vesper_player/vesper_player.dart';
import 'package:vesper_player_external_playback/vesper_player_external_playback.dart';
import 'package:vesper_player_ui/vesper_player_ui.dart' as ui;

import 'example_player_helpers.dart';
import 'example_player_models.dart';
import '../hdr_evidence/hdr_evidence_capture.dart';

part 'example_player_resilience_section.dart';
part 'example_player_plugin_sections.dart';
part 'example_player_system_section.dart';
part 'example_player_playlist_section.dart';
part 'example_player_shared_widgets.dart';

class ExamplePlayerHeader extends StatelessWidget {
  const ExamplePlayerHeader({
    super.key,
    required this.sourceLabel,
    required this.subtitle,
    required this.palette,
  });

  final String sourceLabel;
  final String subtitle;
  final ExampleHostPalette palette;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: <Widget>[
        Text(
          'Vesper',
          style: theme.textTheme.headlineMedium?.copyWith(
            color: palette.title,
            fontWeight: FontWeight.w900,
            letterSpacing: -1.2,
          ),
        ),
        const SizedBox(height: 8),
        Text(
          sourceLabel,
          style: theme.textTheme.titleSmall?.copyWith(
            color: palette.title,
            fontWeight: FontWeight.w600,
          ),
          maxLines: 1,
          overflow: TextOverflow.ellipsis,
        ),
        const SizedBox(height: 6),
        Text(
          subtitle,
          style: theme.textTheme.bodyMedium?.copyWith(
            color: palette.body,
            height: 1.45,
          ),
          maxLines: 2,
          overflow: TextOverflow.ellipsis,
        ),
      ],
    );
  }
}

class ExampleSourceSection extends StatelessWidget {
  const ExampleSourceSection({
    super.key,
    required this.palette,
    required this.themeMode,
    required this.remoteUrlController,
    required this.localFilesEnabled,
    required this.dashEnabled,
    required this.onThemeModeChange,
    required this.onPickVideo,
    required this.onUseHlsDemo,
    required this.onUseDashDemo,
    required this.onUseLiveDvrAcceptance,
    required this.onOpenRemote,
    this.dashUnavailableMessage,
  });

  final ExampleHostPalette palette;
  final ExampleThemeMode themeMode;
  final TextEditingController remoteUrlController;
  final bool localFilesEnabled;
  final bool dashEnabled;
  final ValueChanged<ExampleThemeMode> onThemeModeChange;
  final VoidCallback onPickVideo;
  final VoidCallback onUseHlsDemo;
  final VoidCallback onUseDashDemo;
  final VoidCallback onUseLiveDvrAcceptance;
  final VoidCallback onOpenRemote;
  final String? dashUnavailableMessage;

  @override
  Widget build(BuildContext context) {
    return Container(
      width: double.infinity,
      padding: const EdgeInsets.all(18),
      decoration: BoxDecoration(
        color: palette.sectionBackground,
        borderRadius: BorderRadius.circular(24),
        border: Border.all(color: palette.sectionStroke),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: <Widget>[
          Text(
            '媒体源',
            style: Theme.of(context).textTheme.titleMedium?.copyWith(
              color: palette.title,
              fontWeight: FontWeight.bold,
            ),
          ),
          const SizedBox(height: 14),
          Text(
            '使用这些演示操作在本地文件、HLS、DASH 和自定义远程 URL 之间切换。',
            style: Theme.of(
              context,
            ).textTheme.bodySmall?.copyWith(color: palette.body),
          ),
          const SizedBox(height: 14),
          SingleChildScrollView(
            scrollDirection: Axis.horizontal,
            child: Row(
              children: <Widget>[
                OutlinedButton(
                  onPressed: localFilesEnabled ? onPickVideo : null,
                  child: const Text('选择视频'),
                ),
                const SizedBox(width: 10),
                OutlinedButton(
                  onPressed: onUseHlsDemo,
                  child: const Text('HLS 演示'),
                ),
                const SizedBox(width: 10),
                OutlinedButton(
                  onPressed: onUseLiveDvrAcceptance,
                  child: const Text('Live DVR 验收'),
                ),
                const SizedBox(width: 10),
                OutlinedButton(
                  onPressed: dashEnabled ? onUseDashDemo : null,
                  child: const Text('DASH 演示'),
                ),
              ],
            ),
          ),
          if (dashUnavailableMessage != null) ...<Widget>[
            const SizedBox(height: 10),
            Text(
              dashUnavailableMessage!,
              style: Theme.of(
                context,
              ).textTheme.bodySmall?.copyWith(color: palette.body),
            ),
          ],
          const SizedBox(height: 14),
          TextField(
            controller: remoteUrlController,
            keyboardType: TextInputType.url,
            maxLines: 1,
            decoration: const InputDecoration(labelText: '远程流 URL'),
          ),
          const SizedBox(height: 14),
          Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: <Widget>[
              Text(
                '主题',
                style: Theme.of(context).textTheme.labelLarge?.copyWith(
                  color: palette.title,
                  fontWeight: FontWeight.w600,
                ),
              ),
              const SizedBox(height: 10),
              SingleChildScrollView(
                scrollDirection: Axis.horizontal,
                child: Row(
                  children: <Widget>[
                    ExampleThemeModeChip(
                      icon: Icons.brightness_auto_rounded,
                      label: ExampleThemeMode.system.title,
                      selected: themeMode == ExampleThemeMode.system,
                      palette: palette,
                      onTap: () => onThemeModeChange(ExampleThemeMode.system),
                    ),
                    const SizedBox(width: 10),
                    ExampleThemeModeChip(
                      icon: Icons.light_mode_rounded,
                      label: ExampleThemeMode.light.title,
                      selected: themeMode == ExampleThemeMode.light,
                      palette: palette,
                      onTap: () => onThemeModeChange(ExampleThemeMode.light),
                    ),
                    const SizedBox(width: 10),
                    ExampleThemeModeChip(
                      icon: Icons.dark_mode_rounded,
                      label: ExampleThemeMode.dark.title,
                      selected: themeMode == ExampleThemeMode.dark,
                      palette: palette,
                      onTap: () => onThemeModeChange(ExampleThemeMode.dark),
                    ),
                  ],
                ),
              ),
            ],
          ),
          const SizedBox(height: 14),
          FilledButton(
            onPressed: onOpenRemote,
            style: FilledButton.styleFrom(
              backgroundColor: palette.primaryAction,
              foregroundColor: Colors.white,
            ),
            child: const Text('打开远程 URL'),
          ),
        ],
      ),
    );
  }
}
